File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode* result = nullptr;
        for(auto i : lists){
            if(i == nullptr) continue;
            result = mergeList(result, i);
        }
        return result;
    }

    ListNode* mergeList(ListNode* a, ListNode* b){
        ListNode* result;


        if(a == nullptr || a -> val > b -> val){
            result = b;
            b = b -> next;
        } else {
            result = a;
            a = a -> next;
        }

        ListNode* resultHead = result;

        while(!(a == nullptr && b == nullptr)){
            if(a == nullptr || (b != nullptr && a->val > b->val)){
                result -> next = b;
                b = b -> next;
                result = result -> next;
            } else {
                result -> next = a;
                a = a -> next;
                result = result -> next;
            }
        }

        return resultHead;
    }
};

File: Data Structures & Algorithms/add-two-numbers/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode], c = 0) -> Optional[ListNode]:
        if l1 is None and l2 is None and c==0:
            return None
        
        value = c
        if l1 is not None:
            value += l1.val
            l1 = l1.next
        if l2 is not None: 
            value += l2.val
            l2 = l2.next


        res = ListNode(val = value % 10)

        res.next = self.addTwoNumbers(l1, l2, value // 10)

        return res